nova/quota.py

```diff
@@ -1075,45 +1088,77 @@ def limit_check_project_and_user(self, context, project_values=None,
-            user_values=user_values, project_id=project_id,
-            user_id=user_id)
-
-    def update_resources_from_quota_classes(self):
-        """Update the _resources dict to contain all current resources
-
-        This should be called whenever there's a new dynamic resource added
-        i.e. a resource in the "flavors" quota class.
-        """
-        # NOTE(jkulik): We cannot call this method when getting an API request
-        # to update a quota-class, because then only one of the nova-api pods
-        # would be updated and the others would not - as only a single pod is
-        # handling the request.
-        flavor_quotas = objects.Quotas.get_all_class_by_name(
-            nova_context.get_admin_context(), 'flavors')
-        del flavor_quotas['class_name']
-
-        resources = {}
-        for key, value in flavor_quotas.items():
-            resources[key] = CountableResource(key,
-                _instances_cores_ram_count, default=value)
-
-        resources.update(self._original_resources)
-
-        self._resources = resources
+            context, self._resources, project_values=project_values,
+            user_values=user_values, project_id=project_id, user_id=user_id)
 
     @property
     def resources(self):
-        if self._resources is None:
-            self.update_resources_from_quota_classes()
-        return self._resources
+        return sorted(self._resources.keys())
 
     def get_reserved(self):
         if isinstance(self._driver, NoopQuotaDriver):
             return -1
         return 0
 
 
+class SAPQuotaEngine(QuotaEngine):
+    """SAP specific quota engine
+
+    We extend the resources dynamically based on information in flavors'
+    extra_specs. Through that, we create instance-type quota of the form
+    "instances_<flavor name>" and also quota for different hardware versions
+    for cores/ram of the form "cores_<hw version>"/"ram_<hw version>".
+
+    Since an update to the DB should update our view, we have to re-build the
+    resources regularly. Therefore, we use a cache. Until the cache expires,
+    the resources can be different in different processes.
+    """
+    def __init__(self, resources):
+        # TODO(jkulik) make expiration time configurable
+        self._cache = cache.CacheClient(
+            cache._get_custom_cache_region(expiration_time=600,
+                                           backend='oslo_cache.dict'))
+        super().__init__(resources)
+
+    @property
+    def _resources(self):
+        resources = self._cache.get('resources')
+        if resources is None:
+            resources = {}
+            self.update_resources_from_flavors(resources)
+            resources.update(self._original_resources)
+            self._cache.set('resources', resources)
+        return resources
+
+    @_resources.setter
+    def _resources(self, resources):
+        self._original_resources = resources
+        self._cache.delete('resources')
```

Comment:
If we should decide to use an external cache, then I think it would be better to remove this invalidation. The cache may very well be still valid then.